pvlib/way_faster.py

```diff
@@ -0,0 +1,240 @@
+
+    [1] http://www.pveducation.org/pvcdrom/open-circuit-voltage
+    """
+
+    return nNsVth * np.log(photocurrent / saturation_current + 1.0)
+
+
+def bishop88(vd, photocurrent, saturation_current, resistance_series,
+             resistance_shunt, nNsVth, gradients=False):
+    """
+    Explicit calculation single-diode-model (SDM) currents and voltages using
+    diode junction voltages [1].
+
+    [1] "Computer simulation of the effects of electrical mismatches in
+    photovoltaic cell interconnection circuits" JW Bishop, Solar Cell (1988)
+    https://doi.org/10.1016/0379-6787(88)90059-2
+
+    :param numeric vd: diode voltages [V]
+    :param numeric photocurrent: photo-generated current [A]
+    :param numeric saturation_current: diode one reverse saturation current [A]
+    :param numeric resistance_series: series resitance [ohms]
+    :param numeric resistance_shunt: shunt resitance [ohms]
+    :param numeric nNsVth: product of thermal voltage ``Vth`` [V], diode
+        ideality factor ``n``, and number of series cells ``Ns``
+    :param bool gradients: default returns only i, v, and p, returns gradients
+        if true
+    :returns: tuple containing currents [A], voltages [V], power [W],
+        gradient ``di/dvd``, gradient ``dv/dvd``, gradient ``di/dv``,
+        gradient ``dp/dv``, and gradient ``d2p/dv/dvd``
+    """
+    a = np.exp(vd / nNsVth)
+    b = 1.0 / resistance_shunt
+    i = photocurrent - saturation_current * (a - 1.0) - vd * b
+    v = vd - i * resistance_series
+    retval = (i, v, i*v)
+    if gradients:
+        c = saturation_current * a / nNsVth
+        grad_i = - c - b  # di/dvd
+        grad_v = 1.0 - grad_i * resistance_series  # dv/dvd
+        # dp/dv = d(iv)/dv = v * di/dv + i
+        grad = grad_i / grad_v  # di/dv
+        grad_p = v * grad + i  # dp/dv
+        grad2i = -c / nNsVth  # d2i/dvd
+        grad2v = -grad2i * resistance_series  # d2v/dvd
+        grad2p = (
+            grad_v * grad + v * (grad2i/grad_v - grad_i*grad2v/grad_v**2) + grad_i
+        )  # d2p/dv/dvd
+        retval += (grad_i, grad_v, grad, grad_p, grad2p)
+    return retval
+
+
+def slow_i_from_v(v, photocurrent, saturation_current, resistance_series,
+                  resistance_shunt, nNsVth):
+    """
+    This is a slow but reliable way to find current given any voltage.
+    """
+    if brentq is NotImplemented:
+        raise ImportError('This function requires scipy')
+    # collect args
+    args = (photocurrent, saturation_current, resistance_series,
+            resistance_shunt, nNsVth)
+    # first bound the search using voc
+    voc_est = est_voc(photocurrent, saturation_current, nNsVth)
+    vd = brentq(lambda x, *a: v - bishop88(x, *a)[1], 0.0, voc_est, args)
+    return bishop88(vd, *args)[0]
+
+
+def fast_i_from_v(v, photocurrent, saturation_current, resistance_series,
+                  resistance_shunt, nNsVth):
+    """
+    This is a fast but unreliable way to find current given any voltage.
```

Comment:
I'm not sure if "unreliable" is the best choice of words here. It's more of an "unchecked" result.

Reply:
✔️